feat(app-cmds): cog_command_error for application commands

ToasterUwU opened the pull request:
Fixes #825

## Summary

This adds `cog_application_command_error` which is the application command equivalent of `cog_command_error`.
<!-- What is this pull request for? Does it fix any issues? -->

## Checklist

<!-- Put an x inside [ ] to check it, like so: [x] -->

- [X] If code changes were made then they have been tested.
  - [X] I have updated the documentation to reflect the changes.
  - [X] I have run `task pyright` and fixed the relevant issues.
- [X] This PR fixes an issue.
- [X] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

ToasterUwU commented:
have to add doc stuff, will do that in a moment

ToasterUwU commented:
Ok, so this should be working fine (i tested it) and should be documented completly.

HOWEVER, please check this super carefully. This is the first time that i added code to nextcord, not just changed it or updated the docs. 

So check carefully and give me feedback. Its possible i did something really wrong or stupid, because im not familiar with the codes structure yet.

Skelmis commented on nextcord/application_command.py:
```diff
@@ -510,6 +515,13 @@ def _read_application_commands(self) -> None:
                     value.from_callback(value.callback)
                     self.__cog_application_commands__.append(value)
 
+    def has_error_handler(self) -> bool:
+        """:class:`bool`: Checks whether the cog has an error handler.
+
+        .. versionadded:: 2.4
+        """
+        return not hasattr(self.cog_application_command_error.__func__, "__cog_special_method__")
```
I'd say this naming + documentation is slightly misleading as it also provides the assumption the check handles prefix commands also.

I propose renaming it to something like `has_application_command_error_handler` and the documentation accordingly

Skelmis commented on nextcord/application_command.py:
```diff
@@ -722,6 +761,12 @@ def from_callback(
         if self.name is None:
             self.name = self.callback.__name__
 
+        if self.error_callback is None:
+            if self.parent_cog:
+                self.error_callback = self.parent_cog._get_overridden_method(
+                    self.parent_cog.cog_application_command_error
+                )
+
```
May I ask what purpose this serves here?

None of the values check / used appear to be passed into this method so isn't this just the same as calling this within `__init__`?

ToasterUwU replied:
It is required, because the init method doesn't get a parent_cog object with it, if the slash_command decorator is used.
But when the ClientCog object is initialized, it goes trough all the application commands it contains and updates them with its own object.

So this is required to make the decorator way of adding slash commands work.

Skelmis replied:
Sweet, thanks

Skelmis commented on nextcord/client.py:
```diff
@@ -574,10 +574,12 @@ async def on_application_command_error(
-        # cog = context.cog
-        # if cog and cog.has_error_handler():
-        #     return
+        if interaction.application_command == None:
```
```suggestion
        if interaction.application_command is None:
```

Skelmis commented on nextcord/application_command.py:
```diff
@@ -917,11 +962,10 @@ async def invoke_error(self, interaction: Interaction, error: Exception) -> None
+        if (
+            self.has_error_handler() and self.error_callback is not None
+        ):  # TODO Why is this here? self.has_error_handler does the same as the second check
+            await self.error_callback(interaction, error)
```
How does this handle passing `self` to callbacks if required? Just seems to be something from the removed code, so I'm just curious if that is still applicable here or not

ToasterUwU replied:
It automatically does. The code I removed didn't work.
It passed self automatically, than another self via the object and than the rest args.
Which gave me the error of having 4 args when 3 were exspected.

The code I replaced it with works.

Skelmis reviewed (changes requested):
I haven't tested this yet, will do after getting some more context from my current review items

EmreTech commented on nextcord/application_command.py:
```diff
@@ -722,6 +761,12 @@ def from_callback(
         if self.name is None:
             self.name = self.callback.__name__
 
+        if self.error_callback is None:
+            if self.parent_cog:
```
Couldn't you combine these two if statements together with an and statement?

ToasterUwU replied:
Done

EmreTech commented on nextcord/application_command.py:
```diff
@@ -917,11 +962,10 @@ async def invoke_error(self, interaction: Interaction, error: Exception) -> None
         """
-        if self.has_error_handler() and self.error_callback is not None:
-            if self.parent_cog:
-                await self.error_callback(self.parent_cog, interaction, error)
```
Why isn't the parent cog supplied as a parameter anymore?

ToasterUwU replied:
Take a look at the [last disscussion with Skelmis](https://github.com/nextcord/nextcord/pull/989#discussion_r1103737943) on this exact same thing.

ooliver1 commented on nextcord/application_command.py:
```diff
@@ -510,6 +515,13 @@ def _read_application_commands(self) -> None:
+    def has_application_command_error_handler(self) -> bool:
+        """:class:`bool`: Checks whether the cog has an error handler for application commands.
+
+        .. versionadded:: 2.4
```
```suggestion
        .. versionadded:: 2.5
```

ooliver1 commented on nextcord/application_command.py:
```diff
@@ -536,6 +548,27 @@ def cog_application_command_check(self, interaction: Interaction) -> bool:
+    async def cog_application_command_error(
+        self, interaction: Interaction, error: ApplicationError
+    ) -> None:
+        """A special method that is called whenever an error
+        is dispatched inside this cog.
```
```suggestion
        """A special method that is called whenever an error is dispatched inside this cog.
```
Summaries should be on the first line.

Skelmis commented on nextcord/application_command.py:
```diff
@@ -510,6 +515,13 @@ def _read_application_commands(self) -> None:
+    def has_application_command_error_handler(self) -> bool:
+        """:class:`bool`: Checks whether the cog has an error handler for application commands.
+
+        .. versionadded:: 2.5
```
```suggestion
        .. versionadded:: 2.6
```

Skelmis commented on nextcord/application_command.py:
```diff
@@ -536,6 +548,28 @@ def cog_application_command_check(self, interaction: Interaction) -> bool:
+
+        This **must** be a coroutine.
+
+        .. versionadded:: 2.5
```
```suggestion
        .. versionadded:: 2.6
```

Skelmis commented on nextcord/client.py:
```diff
@@ -574,10 +574,12 @@ async def on_application_command_error(
+        if interaction.application_command is None:
+            return  # Not supposed to ever happen
+
+        cog = interaction.application_command.parent_cog
+        if cog and cog.has_application_command_error_handler():
+            return
```
Sound logic, but wouldn't this be better suited to living in `application_command.has_error_handler`?

ToasterUwU replied:
i simply copied the structure from the prefix commands and all that. Thats how it works there, so i think we should keep it consistend and keep it with slash commands like this as well.

teaishealthy commented on nextcord/application_command.py:
```diff
@@ -917,11 +962,10 @@ async def invoke_error(self, interaction: Interaction, error: Exception) -> None
-                await self.error_callback(self.parent_cog, interaction, error)
-            else:
-                await self.error_callback(interaction, error)
+        if (
+            self.has_error_handler() and self.error_callback is not None
+        ):  # TODO Why is this here? self.has_error_handler does the same as the second check
```
```suggestion
        if self.has_error_handler()
```

teaishealthy commented on nextcord/client.py:
```diff
@@ -574,10 +574,12 @@ async def on_application_command_error(
             return
 
-        # TODO implement cog error handling
-        # cog = context.cog
-        # if cog and cog.has_error_handler():
-        #     return
+        if interaction.application_command is None:
+            return  # Not supposed to ever happen
```
```suggestion
        if interaction.application_command is None:
            return  # Not supposed to ever happen

        if interaction.application_command.has_error_handler():
            return
```

alentoghostflame reviewed (approved):
Haven't tested, but LGTM! If you could merge from master and double check that everything works still, that would be sick.

ToasterUwU commented:
> Haven't tested, but LGTM! If you could merge from master and double check that everything works still, that would be sick.

I dont have the time to write a test and test it again, but I checked if anything has changed that is used by this addition, and non of the code I changed or use for the new stuff has been changed in any of the commits I just merged into the branch. So unless there is something that evades all logic, it should still work just as well. @alentoghostflame

ToasterUwU commented on nextcord/application_command.py:
```diff
@@ -947,11 +992,8 @@ async def invoke_error(self, interaction: Interaction, error: Exception) -> None
-                await self.error_callback(self.parent_cog, interaction, error)
-            else:
-                await self.error_callback(interaction, error)
+        if self.has_error_handler():
```
At that point, why even have the function to check it? I mean, if we never use it.

EmreTech replied:
Useful shorthand for users? Just because it's defined doesn't mean it has to be used by the underlying internal code.

ToasterUwU replied:
I don't think this is a good idea to change. That would mean there is not one spot to change if it ever needs a change, but multiple.
And that, at least for me, is a bigger issue then pyright not being able to put a function with one line into context.

EmreTech replied:
> I don't think this is a good idea to change. That would mean there is not one spot to change if it ever needs a change, but multiple.

`has_error_callback` is currently only referenced once in `application_command.py`. With this suggestion, if error callbacks were changed in some way (maybe to allow multiple or something), it would be a small change. It's really not a big deal in this case.

If you're still not convinced, you can alternatively add this under the if statement:
```python
if TYPE_CHECKING:
    error_callback = cast(Callable, self.error_callback)

await error_callback(interaction, error)
```

ToasterUwU commented on nextcord/client.py:
```diff
@@ -578,13 +578,15 @@ async def on_application_command_error(
         This only fires if you do not specify any listeners for command error.
         """
-        if interaction.application_command and interaction.application_command.has_error_handler():
+        if interaction.application_command is None:
+            return  # Not supposed to ever happen
+
+        if interaction.application_command.has_error_handler():
             return
```
Don't understand the question

EmreTech replied:
These if statements used to be combined together with the `and` operator. Why did you change that?

ToasterUwU replied:
I don't remember anymore, but I think it was because I was fiddling around with it, changed it a lot, and it just ended up like this without me giving it much more thought.

Or I did for the sake of readability, not sure after all this time

EmreTech commented on nextcord/application_command.py:
```diff
@@ -512,6 +517,13 @@ def _read_application_commands(self) -> None:
+    def has_application_command_error_handler(self) -> bool:
+        """:class:`bool`: Checks whether the cog has an error handler for application commands.
+
+        .. versionadded:: 2.6
```
```suggestion
        .. versionadded:: 3.0
```

EmreTech commented on nextcord/application_command.py:
```diff
@@ -538,6 +550,28 @@ def cog_application_command_check(self, interaction: Interaction) -> bool:
+
+        This **must** be a coroutine.
+
+        .. versionadded:: 2.6
```
```suggestion
        .. versionadded:: 3.0
```

EmreTech reviewed (changes requested):
This PR was not considered for 2.6.

ToasterUwU commented:
> This PR was not considered for 2.6.

I mean.. it was.. just like it was for 2.5

It just got moved back twice.

EmreTech reviewed (approved):
LGTM, haven't tested

Skelmis reviewed (changes requested):
The following simplistic bot errors, please fix.
Main: https://paste.nextcord.dev/?id=17002155473974
Cog: https://paste.nextcord.dev/?id=1700215573244250
Error: https://paste.nextcord.dev/?id=1700215592682863

ToasterUwU commented:
Found the issue and fixed it, Because fo the Type checking check, only when type checking there actually will be a error_callback assigned. So tests pass, running it doesn't work lol

Anyways, stupid oversight, fixed now.

Skelmis reviewed (approved):
Tested, lgtm
